From 1be2c702275fee3629cae60c1b50ea9bde89e164 Mon Sep 17 00:00:00 2001
From: Chen Hanxiao <chenhanxiao@cn.fujitsu.com>
Date: Fri, 10 Jan 2014 17:37:55 +0800
Subject: [PATCH 1/2] virt-manager: add support for showing panic notifier
 device

Signed-off-by: Chen Hanxiao <chenhanxiao@cn.fujitsu.com>
(cherry picked from commit a90a375bda19fca19aed591d311774ae992b3beb)

Closes: https://bugzilla.redhat.com/show_bug.cgi?id=996517

Conflicts:
	src/virtManager/details.py
	src/virtManager/domain.py
	ui/details.ui
---
 src/virtManager/details.py |  19 ++++-
 src/virtManager/domain.py  |   3 +
 src/vmm-details.glade      | 190 +++++++++++++++++++++++++++++++++++++++++++++
 3 files changed, 211 insertions(+), 1 deletion(-)

diff --git a/src/virtManager/details.py b/src/virtManager/details.py
index 139dcb4..0e5b274 100644
--- a/src/virtManager/details.py
+++ b/src/virtManager/details.py
@@ -112,13 +112,14 @@ HW_LIST_TYPE_CONTROLLER = 14
 HW_LIST_TYPE_FILESYSTEM = 15
 HW_LIST_TYPE_SMARTCARD = 16
 HW_LIST_TYPE_REDIRDEV = 17
+HW_LIST_TYPE_PANIC = 18
 
 remove_pages = [HW_LIST_TYPE_NIC, HW_LIST_TYPE_INPUT,
                 HW_LIST_TYPE_GRAPHICS, HW_LIST_TYPE_SOUND, HW_LIST_TYPE_CHAR,
                 HW_LIST_TYPE_HOSTDEV, HW_LIST_TYPE_DISK, HW_LIST_TYPE_VIDEO,
                 HW_LIST_TYPE_WATCHDOG, HW_LIST_TYPE_CONTROLLER,
                 HW_LIST_TYPE_FILESYSTEM, HW_LIST_TYPE_SMARTCARD,
-                HW_LIST_TYPE_REDIRDEV]
+                HW_LIST_TYPE_REDIRDEV, HW_LIST_TYPE_PANIC]
 
 # Boot device columns
 BOOT_DEV_TYPE = 0
@@ -1265,6 +1266,8 @@ class vmmDetails(vmmGObjectUI):
                 self.refresh_smartcard_page()
             elif pagetype == HW_LIST_TYPE_REDIRDEV:
                 self.refresh_redir_page()
+            elif pagetype == HW_LIST_TYPE_PANIC:
+                self.refresh_panic_page()
             else:
                 pagetype = -1
         except Exception, e:
@@ -2990,6 +2993,15 @@ class vmmDetails(vmmGObjectUI):
         self.widget("redir-type-label").set_text(rd.type)
         self.widget("redir-type-combo").hide()
 
+    def refresh_panic_page(self):
+        dev = self.get_hw_selection(HW_LIST_COL_DEVICE)
+        if not dev:
+            return
+
+        type = virtinst.VirtualPanicDevice.get_pretty_type(dev.type)
+        self.widget("panic-type").set_text(type or "-")
+        self.widget("panic-iobase").set_text(dev.iobase or "-")
+
     def refresh_char_page(self):
         chardev = self.get_hw_selection(HW_LIST_COL_DEVICE)
         if not chardev:
@@ -3380,6 +3392,11 @@ class vmmDetails(vmmGObjectUI):
             update_hwlist(HW_LIST_TYPE_SMARTCARD, sc,
                           _("Smartcard"), "device_serial")
 
+        # Populate list of Panic devices
+        for panic in self.vm.get_panic_devices():
+            update_hwlist(HW_LIST_TYPE_PANIC, panic,
+                          _("Panic Notifier"), "system-run")
+
         devs = range(len(hw_list_model))
         devs.reverse()
         for i in devs:
diff --git a/src/virtManager/domain.py b/src/virtManager/domain.py
index ecc6302..6cfc50a 100644
--- a/src/virtManager/domain.py
+++ b/src/virtManager/domain.py
@@ -52,6 +52,7 @@ def compare_device(origdev, newdev, idx):
         "filesystem" : ["target" , "vmmindex"],
         "smartcard" : ["mode" , "vmmindex"],
         "redirdev" : ["bus" , "type", "vmmindex"],
+        "panic"       : ["type" , "vmmindex"],
     }
 
     if id(origdev) == id(newdev):
@@ -1010,6 +1011,8 @@ class vmmDomain(vmmLibvirtObject):
         return self._build_device_list("smartcard")
     def get_redirdev_devices(self):
         return self._build_device_list("redirdev")
+    def get_panic_devices(self):
+        return self._build_device_list("panic")
 
     def get_disk_devices(self, refresh_if_necc=True, inactive=False):
         devs = self._build_device_list("disk", refresh_if_necc, inactive)
diff --git a/src/vmm-details.glade b/src/vmm-details.glade
index 5b06359..3f8380a 100644
--- a/src/vmm-details.glade
+++ b/src/vmm-details.glade
@@ -6104,6 +6104,196 @@ I/O:</property>
                               </packing>
                             </child>
 
+                            <child>
+                              <widget class="GtkFrame" id="frame-panic">
+                                <property name="visible">True</property>
+                                <property name="label_xalign">0</property>
+                                <property name="label_yalign">0.5</property>
+                                <property name="shadow_type">none</property>
+
+                                <child>
+                                  <widget class="GtkAlignment" id="alignment-panic">
+                                    <property name="visible">True</property>
+                                    <property name="xalign">0.5</property>
+                                    <property name="yalign">0.5</property>
+                                    <property name="xscale">1</property>
+                                    <property name="yscale">1</property>
+                                    <property name="top_padding">3</property>
+                                    <property name="bottom_padding">0</property>
+                                    <property name="left_padding">12</property>
+                                    <property name="right_padding">0</property>
+
+                                    <child>
+				      <widget class="GtkTable" id="table-panic">
+				        <property name="border_width">3</property>
+				        <property name="visible">True</property>
+				        <property name="n_rows">2</property>
+				        <property name="n_columns">2</property>
+				        <property name="homogeneous">False</property>
+				        <property name="row_spacing">4</property>
+				        <property name="column_spacing">8</property>
+
+				        <child>
+					  <widget class="GtkLabel" id="label-panic-type">
+					    <property name="visible">True</property>
+					    <property name="label" translatable="yes">T_ype:</property>
+					    <property name="use_underline">True</property>
+					    <property name="use_markup">False</property>
+					    <property name="justify">GTK_JUSTIFY_LEFT</property>
+					    <property name="wrap">False</property>
+					    <property name="selectable">False</property>
+					    <property name="xalign">0</property>
+					    <property name="yalign">0.5</property>
+					    <property name="xpad">0</property>
+					    <property name="ypad">0</property>
+					    <property name="mnemonic_widget">panic-type</property>
+					    <property name="ellipsize">PANGO_ELLIPSIZE_NONE</property>
+					    <property name="width_chars">-1</property>
+					    <property name="single_line_mode">False</property>
+					    <property name="angle">0</property>
+					  </widget>
+					  <packing>
+					    <property name="left_attach">0</property>
+					    <property name="right_attach">1</property>
+					    <property name="top_attach">0</property>
+					    <property name="bottom_attach">1</property>
+					    <property name="x_options"></property>
+					    <property name="y_options"></property>
+					  </packing>
+				        </child>
+
+				        <child>
+					  <widget class="GtkLabel" id="panic-type">
+					    <property name="visible">True</property>
+					    <property name="can_focus">True</property>
+					    <property name="label">panic type</property>
+					    <property name="use_underline">False</property>
+					    <property name="use_markup">False</property>
+					    <property name="justify">GTK_JUSTIFY_LEFT</property>
+					    <property name="wrap">False</property>
+					    <property name="selectable">True</property>
+					    <property name="xalign">0</property>
+					    <property name="yalign">0.5</property>
+					    <property name="xpad">1</property>
+					    <property name="ypad">0</property>
+					    <property name="ellipsize">PANGO_ELLIPSIZE_NONE</property>
+					    <property name="width_chars">-1</property>
+					    <property name="single_line_mode">False</property>
+					    <property name="angle">0</property>
+					  </widget>
+					  <packing>
+					    <property name="left_attach">1</property>
+					    <property name="right_attach">2</property>
+					    <property name="top_attach">0</property>
+					    <property name="bottom_attach">1</property>
+					    <property name="x_options">fill</property>
+					    <property name="y_options"></property>
+					  </packing>
+				        </child>
+
+				        <child>
+					  <widget class="GtkLabel" id="label520">
+					    <property name="visible">True</property>
+					    <property name="label" translatable="yes">IO Base:</property>
+					    <property name="use_underline">True</property>
+					    <property name="use_markup">False</property>
+					    <property name="justify">GTK_JUSTIFY_LEFT</property>
+					    <property name="wrap">False</property>
+					    <property name="selectable">False</property>
+					    <property name="xalign">0</property>
+					    <property name="yalign">0.5</property>
+					    <property name="xpad">0</property>
+					    <property name="ypad">0</property>
+					    <property name="mnemonic_widget">panic-type</property>
+					    <property name="ellipsize">PANGO_ELLIPSIZE_NONE</property>
+					    <property name="width_chars">-1</property>
+					    <property name="single_line_mode">False</property>
+					    <property name="angle">0</property>
+					  </widget>
+					  <packing>
+					    <property name="left_attach">0</property>
+					    <property name="right_attach">1</property>
+					    <property name="top_attach">1</property>
+					    <property name="bottom_attach">2</property>
+					    <property name="x_options">fill</property>
+					    <property name="y_options"></property>
+					  </packing>
+				        </child>
+
+				        <child>
+					  <widget class="GtkLabel" id="panic-iobase">
+					    <property name="visible">True</property>
+					    <property name="can_focus">True</property>
+					    <property name="label">panic iobase</property>
+					    <property name="use_underline">False</property>
+					    <property name="use_markup">False</property>
+					    <property name="justify">GTK_JUSTIFY_LEFT</property>
+					    <property name="wrap">False</property>
+					    <property name="selectable">True</property>
+					    <property name="xalign">0</property>
+					    <property name="yalign">0.5</property>
+					    <property name="xpad">1</property>
+					    <property name="ypad">0</property>
+					    <property name="ellipsize">PANGO_ELLIPSIZE_NONE</property>
+					    <property name="width_chars">-1</property>
+					    <property name="single_line_mode">False</property>
+					    <property name="angle">0</property>
+					  </widget>
+					  <packing>
+					    <property name="left_attach">1</property>
+					    <property name="right_attach">2</property>
+					    <property name="top_attach">1</property>
+					    <property name="bottom_attach">2</property>
+					    <property name="x_options">fill</property>
+					    <property name="y_options"></property>
+					  </packing>
+				        </child>
+				      </widget>
+	                            </child>
+                                  </widget>
+                                </child>
+
+                                <child>
+                                  <widget class="GtkLabel" id="panic-title">
+                                    <property name="visible">True</property>
+                                    <property name="label" translatable="yes">&lt;b&gt;Panic device&lt;/b&gt;</property>
+                                    <property name="use_underline">False</property>
+                                    <property name="use_markup">True</property>
+                                    <property name="justify">GTK_JUSTIFY_LEFT</property>
+                                    <property name="wrap">False</property>
+                                    <property name="selectable">False</property>
+                                    <property name="xalign">0.5</property>
+                                    <property name="yalign">0.5</property>
+                                    <property name="xpad">0</property>
+                                    <property name="ypad">0</property>
+                                    <property name="ellipsize">PANGO_ELLIPSIZE_NONE</property>
+                                    <property name="width_chars">-1</property>
+                                    <property name="single_line_mode">False</property>
+                                    <property name="angle">0</property>
+                                  </widget>
+                                  <packing>
+                                    <property name="type">label_item</property>
+                                  </packing>
+                                </child>
+                              </widget>
+                              <packing>
+                                <property name="position">18</property>
+                                <property name="tab_expand">False</property>
+                                <property name="tab_fill">True</property>
+                              </packing>
+                            </child>
+
+                            <child>
+                              <widget class="GtkLabel" id="label515">
+                                <property name="visible">True</property>
+                                <property name="label">panic</property>
+                              </widget>
+                              <packing>
+                                <property name="position">17</property>
+                                <property name="tab_fill">False</property>
+                                <property name="type">tab</property>
+                              </packing>
+                            </child>
                           </widget>
                         </child>
                       </widget>
-- 
1.9.3

